fex: fix tests + minimize set of cloned submodules

kuruczgy opened the pull request:
See the commit messages for more details.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

kuruczgy committed:
fex: minimize set of cloned submodules

This change also removes the submodules that contained prebuilt (test)
binaries.

kuruczgy committed:
fex: fix and enable tests

This will still fail on Apple Silicon devices using 16K page size
kernels, but such systems are not supported by upstream, so there is
nothing we can do about that.

kuruczgy commented on pkgs/by-name/fe/fex/package.nix:
```diff
@@ -42,25 +69,66 @@ llvmPackages.stdenv.mkDerivation (finalAttrs: rec {
-  doCheck = false; # broken on Apple silicon computers
+
+  # Unsupported on non-4K page size kernels (e.g. Apple Silicon)
+  doCheck = stdenv.buildPlatform.canExecute stdenv.hostPlatform;
```
Unfortunately there is no way to check the kernel page size here, so people wishing to compile this on Apple Silicon will have to use `.override { doCheck = false; }`.

I think this is the right thing to do, having Hydra run the tests is more valuable than being able to build this package without changes on AS. (Most users won't be impacted since they will just pull the package from the cache anyway. And if you are developing the package on AS, a temporary change to disable the tests should not be that big of a deal. I think this is preferable to requiring developers from every other platform to manually enable the tests instead.)

andre4ik3 replied:
```suggestion
  doCheck = stdenv.buildPlatform.canExecute stdenv.hostPlatform;

  nativeCheckInputs = [ nasm ];
```

andre4ik3 reviewed (approved):
Diff LGTM, thank you!

kuruczgy commented:
Tested, runs HL2 fine.

(Also for some reason https://github.com/FEX-Emu/FEX/issues/1647 no longer seems to be an issue, maybe https://github.com/FEX-Emu/FEX/pull/4225 fixed it? Anyway, not relevant for this PR.)

kuruczgy commented:
@Aleksanaa, you helped merge the original fex init PR, would you mind perhaps having a look at this as well? (and https://github.com/NixOS/nixpkgs/pull/374691)

kuruczgy commented:
Rebased. Did not re-test anything since I am on an x86 machine right now but I would be very surprised if only a rebase would break anything...

andre4ik3 reviewed (approved):
Tested in a virtual machine, using a RootFS built with `buildFHSEnv`, compiles and installs fine, all tests pass. Runs Bash fine, unfortunately couldn't test GUI apps as OrbStack doesn't support it. Like you said, it still builds on Apple Silicon machines if `doCheck` is set to false, and I agree that it's better to run the tests than have it compile on technically-unsupported machines :)

kuruczgy commented:
Just FYI, this FEX issue regarding dynamically linked libraries was brought to my attention recently: https://github.com/FEX-Emu/FEX/issues/4301

> At the end of the day we should only dynamically link against PT_INTERP, libc, libgcc_s, libm, and libstdc++.

The FEX people are still looking into this, so it's hard to say what we should do exactly. I was told on their Discord that:

> FWIW the worst that'll happen right now is that you won't be able to safely enable library forwarding for certain 32-bit apps (those apps would crash if you enabled it)

**So for now I would say that this PR can go ahead**, we are not using thunking yet so this should not affect us, but we should look into it more when we consider bringing up any 32-bit thunk libs.

kuruczgy commented:
Rebased onto #380700, please merge that one first.

kuruczgy commented:
Tested as a sanity check, still runs HL2.

@andre4ik3, do you perhaps know of some committer who we could ping about this PR?

andre4ik3 commented:
Unfortunately no, but I put it in nixpkgs review requests on Matrix, hopefully it'll get picked up there.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-already-reviewed/2617/2240

Aleksanaa commented:
@ofborg build fex

kuruczgy commented:
Rebased. I removed the `preCheck` hook that disabled the failing Oryon tests, as now they are disabled upstream: https://github.com/FEX-Emu/FEX/pull/4309

(Tested, still runs HL2)

andre4ik3 commented on pkgs/by-name/fe/fex/package.nix:
```diff
@@ -36,25 +63,55 @@ llvmPackages.stdenv.mkDerivation (finalAttrs: rec {
       ]
     ))
-  ];
+  ] ++ lib.optionals finalAttrs.doCheck [ nasm ];
```
I assumed it was because `nasm` would be a build-phase dependency instead of a check-phase dependency. But it looks like it works both ways.

kuruczgy replied:
> Why not putting this in nativeCheckInputs instead?

I did not use it because I did not know it existed :) Fixed.

> I assumed it was because nasm would be a build-phase dependency instead of a check-phase dependency.

I checked, it has nothing to do with the check phase, it is the exact same `++ optionals doCheck [ ... ]` logic that I was using here: https://github.com/NixOS/nixpkgs/blob/886b3f8376c8383dae7f0a6e9d155dbdec312266/pkgs/stdenv/generic/make-derivation.nix#L338

SuperSandro2000 commented on pkgs/by-name/fe/fex/package.nix:
```diff
@@ -67,6 +126,7 @@ llvmPackages.stdenv.mkDerivation (finalAttrs: rec {
     description = "Fast usermode x86 and x86-64 emulator for Arm64 Linux";
     homepage = "https://fex-emu.com/";
     platforms = [ "aarch64-linux" ];
+    sourceProvenance = [ lib.sourceTypes.fromSource ];
```
```suggestion
```
thats the default

kuruczgy replied:
fixed

SuperSandro2000 commented on pkgs/by-name/fe/fex/package.nix:
```diff
@@ -38,23 +65,55 @@ llvmPackages.stdenv.mkDerivation (finalAttrs: rec {
+    "fexcore_apitests"
+    "emitter_tests"
+
+    # Needs x86 libc headers
+    # "struct_verifier"
+
+    # Needs x86 compiler
+    # "fex_linux_tests_all"
+
+    # Needs thunks to be enabled
+    # "thunkgen_tests"
+
+    # Needs prebuilt binaries
+    # "gcc_target_tests_64"
+    # "gcc_target_tests_32"
+    # "posix_tests"
+    # "gvisor_tests"
```
Why not remove those and add a comment that not everything was added here?

kuruczgy replied:
I wanted to add a rationale for each one I decided to exclude. But I don't have a strong opinion on this, so since you both seem to agree on this, I swapped them to a single more generic comment as you suggested.

SuperSandro2000 commented on pkgs/by-name/fe/fex/package.nix:
```diff
@@ -2,31 +2,58 @@
+
+      # Only fetch required submodules
+      git submodule update --init --depth 1 \
+        External/Vulkan-Headers \
+        External/drm-headers \
+        External/jemalloc \
+        External/jemalloc_glibc \
+        External/robin-map \
+        External/vixl \
+        Source/Common/cpp-optparse \
+        External/Catch2 \
+        External/xbyak
```
Why are we not using our versions of those libraries?

kuruczgy replied:
That is the long term goal, here is the upstream issue tracking this: https://github.com/FEX-Emu/FEX/issues/2996

This PR is a move into this direction, by eliminating the `External/fmt` and `External/xxhash` submodules, and using our own libraries instead.

I believe `catch2` should also be an easy win, and I do plan on looking into swapping it for our own package in a follow up PR. (But I don't want to scope creep this one further, it's already large enough.)

GaetanLepage reviewed (approved):
LGTM

kuruczgy commented:
Gentle ping @SuperSandro2000, please look at my fixes/comments and update your reviews.

andre4ik3 commented:
Made an update PR for the new 2504 version: #396450

It also includes the missing version fix from this PR which I noticed while testing so I put you as co-author. :)

If/when you rebase I'll try to get the PR reviewed by some people on Matrix again. I know it's been like almost 3 months but honestly I still have hope in this getting merged

andre4ik3 commented on pkgs/by-name/fe/fex/package.nix:
```diff
@@ -38,23 +65,42 @@ llvmPackages.stdenv.mkDerivation (finalAttrs: rec {
   strictDeps = true;
-  doCheck = false; # broken on Apple silicon computers
+
+  # Unsupported on non-4K page size kernels (e.g. Apple Silicon)
+  doCheck = stdenv.buildPlatform.canExecute stdenv.hostPlatform;
```
Also can you clarify what this line is meant to do? `doCheck` is false by default, this sets it to true if the `buildPlatform` can execute the `hostPlatform`, which is only the case for native builds (as from my understanding check-phase never runs in cross-compilation anyway). So wouldn't it be simpler to just put `doCheck = true;` ? Or leave it out for the default behavior of no check (to speed up compiling), and only run tests on package updates?

kuruczgy replied:
> buildPlatform can execute the hostPlatform, which is only the case for native builds

`x86_64-linux` can execute `i686-linux` AFAIK.

> as from my understanding [check-phase never runs in cross-compilation anyway](https://nixos.org/manual/nixpkgs/stable/#var-stdenv-doCheck)

The manual is being a bit imprecise here, the exact behavior is this:
https://github.com/NixOS/nixpkgs/blob/59e663d00c2ab3cac67182eb24a1e9a6e2ab3e3a/pkgs/stdenv/generic/make-derivation.nix#L298

So right now, `doCheck = true` and `doCheck = stdenv.buildPlatform.canExecute stdenv.hostPlatform` are exactly equivalent.

Asked about this on matrix in `#dev:nixos.org`, the answer I got for which one to use is to "just do whatever you feel like :)" (I also pinged you if you want to read the whole discussion there.)

Anyway I am also leaning towards `doCheck = true` because "Then can remove the `stdenv` from the arguments as well."

If I do an update of this PR (e.g. a rebase if https://github.com/NixOS/nixpkgs/pull/396450 gets merged first), I can change this.

kuruczgy commented:
Rebased.

> Also, it should probably be squashed into one commit.

Now we are down to only two commits, and they do make sense independently, so I kept them separate.

GaetanLepage reviewed (approved):
LGTM!
